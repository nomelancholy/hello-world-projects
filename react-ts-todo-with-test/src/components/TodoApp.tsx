import React, { useState, useRef } from "react";
import TodoForm from "./TodoForm";
import { TodoModel } from "../model/Todo";
import TodoList from "./TodoList";

const TodoApp = () => {
  const [todos, setTodos] = useState<TodoModel[]>([
    {
      id: 1,
      text: "TDD 배우기",
      done: true,
    },
    {
      id: 2,
      text: "react-testing-library 사용하기",
      done: true,
    },
  ]);

  const [nextId, setNextId] = useState(3);

  const onInsert = (value: string) => {
    setTodos(
      todos.concat({
        id: nextId,
        text: value,
        done: false,
      })
    );

    setNextId(nextId + 1);
  };

  const onToggle = (id: number) => {
    setTodos(
      todos.map((todo) =>
        todo.id === id ? { ...todo, done: !todo.done } : todo
      )
    );
  };

  const onRemove = (id:number) => {
      setTodos(todos.filter(todo => todo.id !== id));
  }

  return (
    <>
      <TodoForm data-testid="helloworld" onInsert={onInsert} />
      <TodoList todos={todos} onToggle={onToggle} onRemove={onRemove} />
    </>
  );
};

export default TodoApp;
