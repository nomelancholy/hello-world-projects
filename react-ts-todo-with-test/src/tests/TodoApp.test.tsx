import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoApp from "../components/TodoApp";

describe("<TodoApp />", () => {
  it("matches snapshot", () => {
    const utils = render(<TodoApp />);
    expect(utils.container).toMatchSnapshot();
  });
  it(`renders TodoForm and TodoList`, () => {
    const { getByText, getByTestId } = render(<TodoApp />);
    getByText("등록");
    getByTestId("TodoList");
  });

  it("renders two defaults todos", () => {
    const { getByText } = render(<TodoApp />);
    getByText("TDD 배우기");
    getByText("react-testing-library 사용하기");
  });

  it('creates new todo', () => {
    const { getByPlaceholderText, getByText } = render(<TodoApp />);
    fireEvent.change(getByPlaceholderText('할 일을 입력하세요'), {
        target: {
            value : '새 항목 추가하기'
        }
    });
    fireEvent.click(getByText('등록'));
    getByText('새 항목 추가하기')
  })

  it('toggles todo', () => {
      const { getByText } = render(<TodoApp />);
      const todoText = getByText('TDD 배우기');
      expect(todoText).toHaveClass('on')
      fireEvent.click(todoText);
      expect(todoText).not.toHaveClass('on')
      fireEvent.click(todoText);
      expect(todoText).toHaveClass('on')
  })

  it('removes todo', () => {
      const { getByText } = render(<TodoApp />);
      const todoText = getByText('TDD 배우기');
      const removeButton = todoText.nextElementSibling!;
      fireEvent.click(removeButton);
      expect(todoText).not.toBeInTheDocument();
  })
});
