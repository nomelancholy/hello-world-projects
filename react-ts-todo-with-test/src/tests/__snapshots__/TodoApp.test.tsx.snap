// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoApp /> matches snapshot 1`] = `
<div>
  <form>
    <input
      placeholder="할 일을 입력하세요"
      value=""
    />
    <button
      type="submit"
    >
      등록
    </button>
  </form>
  <ul
    data-testid="TodoList"
  >
    <li>
      <span
        class="on"
      >
        TDD 배우기
      </span>
      <button>
        삭제
      </button>
    </li>
    <li>
      <span
        class="on"
      >
        react-testing-library 사용하기
      </span>
      <button>
        삭제
      </button>
    </li>
  </ul>
</div>
`;
